File: constrained-device-app/src/test/python/programmingtheiot/part03/integration/connection/MqttClientPerformanceTest.py
#####
# 
# This class is part of the Programming the Internet of Things
# project, and is available via the MIT License, which can be
# found in the LICENSE file at the top level of this repository.
# 
# Copyright (c) 2020 by Andrew D. King
# 

import logging
import unittest

import  time

import programmingtheiot.common.ConfigConst as ConfigConst

from programmingtheiot.cda.connection.MqttClientConnector import MqttClientConnector
from programmingtheiot.common.ConfigUtil import ConfigUtil
from programmingtheiot.common.ResourceNameEnum import ResourceNameEnum
from programmingtheiot.common.DefaultDataMessageListener import DefaultDataMessageListener
from programmingtheiot.data.SensorData import SensorData
from programmingtheiot.data.DataUtil import DataUtil

class MqttClientPerformanceTest(unittest.TestCase):
	"""
	This test case class contains very basic unit tests for
	MqttClientConnector. It should not be considered complete,
	but serve as a starting point for the student implementing
	additional functionality within their Programming the IoT
	environment.
	"""
	NS_IN_MILLIS = 1000000
	
	# NOTE: We'll use only 10,000 requests for MQTT
	MAX_TEST_RUNS = 10000
	
	@classmethod
	def setUpClass(self):
		logging.basicConfig(format = '%(asctime)s:%(module)s:%(levelname)s:%(message)s', level = logging.DEBUG)
		
	def setUp(self):
		self.mqttClient = MqttClientConnector(clientID = 'CDAMqttClientPerformanceTest001')
		pass

	def tearDown(self):
		pass

	@unittest.skip("Ignore for now.")
	def testConnectAndDisconnect(self):
		startTime = time.time_ns()
		
		self.assertTrue(self.mqttClient.connectClient())
		self.assertTrue(self.mqttClient.disconnectClient())
		
		endTime = time.time_ns()
		elapsedMillis = (endTime - startTime) / self.NS_IN_MILLIS
		
		logging.info("Connect and Disconnect: " + str(elapsedMillis) + " ms")
		
	@unittest.skip("Ignore for now.")
	def testPublishQoS0(self):
		self._execTestPublish(self.MAX_TEST_RUNS, 0)

	@unittest.skip("Ignore for now.")
	def testPublishQoS1(self):
		self._execTestPublish(self.MAX_TEST_RUNS, 1)

	@unittest.skip("Ignore for now.")
	def testPublishQoS2(self):
		self._execTestPublish(self.MAX_TEST_RUNS, 2)


	def _execTestPublish(self, maxTestRuns: int, qos: int):
		self.assertTrue(self.mqttClient.connectClient())
		
		sensorData = SensorData()
		payload = DataUtil().sensorDataToJson(sensorData)
		
		startTime = time.time_ns()
		
		for seqNo in range(0, maxTestRuns):
			self.mqttClient.publishMessage(resource = ResourceNameEnum.CDA_SENSOR_MSG_RESOURCE, msg = payload, qos = qos)
			
		endTime = time.time_ns()
		elapsedMillis = (endTime - startTime) / self.NS_IN_MILLIS
		
		self.assertTrue(self.mqttClient.disconnectClient())
		
		logging.info("Publish message - QoS " + str(qos) + " [" + str(maxTestRuns) + "]: " + str(elapsedMillis) + " ms")

if __name__ == "__main__":
	unittest.main()
	

File: gateway-device-app/src/test/java/programmingtheiot/part03/integration/connection/MqttClientPerformanceTest.java
/**
 * 
 * This class is part of the Programming the Internet of Things
 * project, and is available via the MIT License, which can be
 * found in the LICENSE file at the top level of this repository.
 * 
 * Copyright (c) 2020 by Andrew D. King
 */ 

package programmingtheiot.part03.integration.connection;

import static org.junit.Assert.*;

import java.util.logging.Logger;

import org.junit.After;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;

import programmingtheiot.common.ConfigConst;
import programmingtheiot.common.ConfigUtil;
import programmingtheiot.common.DefaultDataMessageListener;
import programmingtheiot.common.IDataMessageListener;
import programmingtheiot.common.ResourceNameEnum;
import programmingtheiot.data.DataUtil;
import programmingtheiot.data.SensorData;
import programmingtheiot.gda.connection.*;

/**
 * This test case class contains very basic integration tests for
 * MqttClientConnectorTest. It should not be considered complete,
 * but serve as a starting point for the student implementing
 * additional functionality within their Programming the IoT
 * environment.
 *
 */
public class MqttClientPerformanceTest
{
	// static
	
	private static final Logger _Logger =
		Logger.getLogger(MqttClientPerformanceTest.class.getName());
	
	
	// member var's
	
	private MqttClientConnector mqttClient = null;
	
	
	// test setup methods
	// test methods
	
	/**
	 * Test method for {@link programmingtheiot.gda.connection.MqttClientConnector#connectClient()}.
	 */
	// NOTE: We'll use only 10,000 requests for MQTT
	public static final int MAX_TEST_RUNS = 10000;
	
	@Before
	public void setUp() throws Exception
	{
		this.mqttClient = new MqttClientConnector();
	}
	//@Ignore
	@Test
	public void testConnectAndDisconnect()
	{
		long startMillis = System.currentTimeMillis();
		
		assertTrue(this.mqttClient.connectClient());
		assertTrue(this.mqttClient.disconnectClient());
		
		long endMillis = System.currentTimeMillis();
		long elapsedMillis = endMillis - startMillis;
		
		_Logger.info("Connect and Disconnect: " + elapsedMillis + " ms");
	}
	@Ignore
	@Test
	public void testPublishQoS0()
	{
		execTestPublish(MAX_TEST_RUNS, 0);
	}
	@Ignore
	@Test
	public void testPublishQoS1()
	{
		execTestPublish(MAX_TEST_RUNS, 1);
	}
	@Ignore
	@Test
	public void testPublishQoS2()
	{
		execTestPublish(MAX_TEST_RUNS, 2);
	}
	
	private void execTestPublish(int maxTestRuns, int qos)
	{
		assertTrue(this.mqttClient.connectClient());
		
		SensorData sensorData = new SensorData();
		
		String payload = DataUtil.getInstance().sensorDataToJson(sensorData);
		
		long startMillis = System.currentTimeMillis();
		
		for (int sequenceNo = 0; sequenceNo < maxTestRuns; sequenceNo++) {
			this.mqttClient.publishMessage(ResourceNameEnum.CDA_MGMT_STATUS_CMD_RESOURCE, payload, qos);
		}
		
		long endMillis = System.currentTimeMillis();
		long elapsedMillis = endMillis - startMillis;
		
		assertTrue(this.mqttClient.disconnectClient());
		
		_Logger.info("Publish message - QoS " + qos + " [" + maxTestRuns + "]: " + elapsedMillis + " ms");
	}
	
}
